File: sdk/python/weaver_sdk/sse.py
from __future__ import annotations

import contextlib
import json
from typing import Any

from .types import SseEvent


def parse_sse_frame(frame: str) -> SseEvent | None:
    text = str(frame or "")
    if not text.strip():
        return None

    event_name = ""
    id_text = ""
    data_lines: list[str] = []

    for raw_line in text.split("\n"):
        line = raw_line.rstrip("\n").rstrip("\r")
        if not line:
            continue
        if line.startswith(":"):
            continue
        if line.startswith("event:"):
            event_name = line[len("event:") :].strip()
            continue
        if line.startswith("id:"):
            id_text = line[len("id:") :].strip()
            continue
        if line.startswith("data:"):
            data_lines.append(line[len("data:") :].lstrip())
            continue

    if not data_lines:
        return None

    data_text = "\n".join(data_lines)
    try:
        parsed: Any = json.loads(data_text)
    except Exception:
        return None

    out: SseEvent = {"data": parsed}
    if event_name:
        out["event"] = event_name
    if id_text:
        with contextlib.suppress(Exception):
            out["id"] = int(id_text)
    return out


def iter_sse_events_from_text(text: str) -> list[SseEvent]:
    """
    Best-effort parse SSE events from an accumulated text buffer.

    This is mostly useful for unit tests and small payloads. Streaming parsing
    should be done incrementally (see WeaverClient.chat_sse()).
    """
    buf = (text or "").replace("\r\n", "\n")
    frames = buf.split("\n\n")
    events: list[SseEvent] = []
    for frame in frames:
        parsed = parse_sse_frame(frame)
        if parsed:
            events.append(parsed)
    return events

File: sdk/typescript/src/sse.ts
import type { SseEvent } from './types.js'

export function parseSseFrame(frame: string): SseEvent | null {
  const text = String(frame ?? '')
  if (!text.trim()) return null

  const lines = text.split('\n').map((l) => l.trimEnd())
  let eventName = ''
  let idText = ''
  const dataLines: string[] = []

  for (const line of lines) {
    if (!line) continue
    if (line.startsWith(':')) continue

    if (line.startsWith('event:')) {
      eventName = line.slice('event:'.length).trim()
      continue
    }

    if (line.startsWith('id:')) {
      idText = line.slice('id:'.length).trim()
      continue
    }

    if (line.startsWith('data:')) {
      dataLines.push(line.slice('data:'.length).trimStart())
      continue
    }
  }

  if (dataLines.length === 0) return null
  const dataText = dataLines.join('\n')

  let parsed: unknown
  try {
    parsed = JSON.parse(dataText) as unknown
  } catch {
    return null
  }

  const out: SseEvent = { data: parsed }
  if (eventName) out.event = eventName
  if (idText) {
    const idNum = Number(idText)
    if (Number.isFinite(idNum)) out.id = idNum
  }

  return out
}

export async function* readSseEvents(response: Response): AsyncGenerator<SseEvent> {
  if (!response.body) return

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    buffer = buffer.replace(/\r\n/g, '\n')

    const frames = buffer.split('\n\n')
    buffer = frames.pop() || ''

    for (const frame of frames) {
      const parsed = parseSseFrame(frame)
      if (parsed) yield parsed
    }
  }

  const tail = buffer.trim()
  if (tail) {
    const parsed = parseSseFrame(tail)
    if (parsed) yield parsed
  }
}

